MCprep_addon/optimize_scene.py

```diff
@@ -218,10 +237,63 @@ def execute(self, context):
 		for mat in bpy.data.materials:
 			matGen = util.nameGeneralize(mat.name)
 			canon, form = generate.get_mc_canonical_name(matGen)
+			mat_type = None
+			homogenous_volumes = 0
+			not_homogenous_volumes = 0
 			if generate.checklist(canon, "reflective"):
 				Glossy += 1
+				mat_type = "reflective"
 			if generate.checklist(canon, "glass"):
 				Transmissive += 1
+				mat_type = "glass"
+
+			if mat.use_nodes:
+				nodes = mat.node_tree.nodes
+				for node in nodes:
+					if node.bl_idname == "ShaderNodeVolumeScatter" or \
+						node.bl_idname == "ShaderNodeVolumeAbsorption" or \
+						node.bl_idname == "ShaderNodeVolumePrincipled":
+						density_socket = node.inputs["Density"] # Grab the density
+						node_name = util.nameGeneralize(node.name).rstrip()  # Get the name (who knew this could be used on nodes?)
+						# Sometimes there may be something linked to the density but it's fine to treat it as a homogeneous volume
+						# This allows the user to control the addon at the node level
+						if not density_socket.is_linked:
+							if node_name == MCPREP_NOT_HOMOGENOUS_VOLUME:
+								not_homogenous_volumes -= 1
+							else:
+								homogenous_volumes += 1
+						else:
+							if node_name == MCPREP_HOMOGENOUS_VOLUME:
+								homogenous_volumes += 1
+							else:
+								not_homogenous_volumes += 1
+
+						ScramblingMultiplier += SCRAMBLING_MULTIPLIER_ADD
+						if ScramblingMultiplier >= CMP_SCRAMBLING_MULTIPLIER: # at this point, it's worthless to keep it enabled 
+							UseScrambling = False
+							PreviewScrambling = False
+							ScramblingMultiplier = 1.0
+					# Not the best check, but better then nothing
+					elif node.bl_idname == "ShaderNodeBsdfPrincipled":
+						if mat_type == "reflective":
+							roughness_socket = node.inputs["Roughness"]
+							if not roughness_socket.is_linked and roughness_socket.default_value >= 0.2:
+								Glossy = Glossy - 1 if Glossy > 1 else Glossy
```

Comment:
nit alt phrasing idea that is easier to read (also note the renaming to lower case... I can also make that change post merge if you don't want to spend time renaming vars):

```
glossy -= 1 if glossy > 1 else 0
```